### PR TITLE
Add dashboard toolbar

### PR DESCRIPTION
**Pull request by kisantia**
This adds dashboard toolbar and moves actions(backup, restore, new query, new notebook) from Tasks widget to toolbar.

Server
image

Database
image

Azure db - no backup and restore actions in toolbar
image

### CODE REVIEW

File: src/sql/media/icons/backup-toolbar.svg [hunk 1]
```diff
@@ -0,0 +1,4 @@
+<svg width="16" height="16" viewBox="0 0 16 16" fill="none" xmlns="http://www.w3.org/2000/svg">
```
Comment (Charles-Gagnon) [inline src/sql/media/icons/backup-toolbar.svg]: Did you reach out to get dark-mode icons as well?

File: src/sql/workbench/contrib/dashboard/browser/core/actions.ts [hunk 2]
```diff
@@ -75,6 +75,116 @@ export class RefreshWidgetAction extends Action {
...
+	) {
+		super(RestoreToolbarAction.ID, RestoreToolbarAction.LABEL, RestoreToolbarAction.CSSCLASS);
+	}
+
+	run(): Promise<boolean> {
+		try {
+			this.restoreFn.apply(this.context);
+			return Promise.resolve(true);
+		} catch (e) {
+			return Promise.resolve(false);
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/dashboard/browser/core/actions.ts]: This error should really be logged somewhere, at least send something to the logger service I'd say. (same for the other actions you added)

File: src/sql/workbench/contrib/dashboard/browser/core/dashboardPage.component.ts [hunk 3]
```diff
@@ -126,6 +151,48 @@ export abstract class DashboardPage extends AngularDisposable implements IConfig
...
 			this.createTabs(tempWidgets);
 		}
+
+		this.createTaskbar(this.taskContainer.nativeElement);
+	}
+
+	protected createTaskbar(parentElement: HTMLElement): void {
+		let taskbarContainer = DOM.append(parentElement, DOM.$('div'));
+		this.taskbar = this._register(new Taskbar(taskbarContainer));
+		this._restoreAction = new RestoreToolbarAction(this.restore, this);
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/dashboard/browser/core/dashboardPage.component.ts]: What if a dashboard doesn't want all these actions (such as the BDC dashboard)? Ideally these actions should be contributed through a more generic means that each dashboard type sets itself.

File: src/sql/workbench/contrib/dashboard/browser/core/dashboardPage.component.ts [hunk 4]
```diff
@@ -126,6 +151,48 @@ export abstract class DashboardPage extends AngularDisposable implements IConfig
...
+		this._editAction = new EditDashboardAction(this.enableEdit, this);
+		this._refreshAction = new RefreshWidgetAction(this.refresh, this);
+		this._manageExtensionsAction = new ManageExtensionsToolbarAction(this.manageExtensions, this);
+		this.setTaskbarContent();
+	}
+
+	protected setTaskbarContent(): void {
+		// Create HTML Elements for the taskbar
+		let separator: HTMLElement = Taskbar.createTaskbarSeparator();
+		separator.style.margin = '0px 15px 0px 0px';
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/dashboard/browser/core/dashboardPage.component.ts]: Why isn't this done in the createTaskbarSeparator? Having random values like this floating around make it difficult to track down if there's issues.

And anyways this should probably go into a CSS anyways

File: src/sql/workbench/contrib/dashboard/browser/core/dashboardPage.component.ts [hunk 5]
```diff
@@ -353,6 +420,23 @@ export abstract class DashboardPage extends AngularDisposable implements IConfig
...
+		this._instantiationService.invokeFunction(openNewQuery);
+	}
+
+	public newNotebook(): Promise<void> {
+		return this.commandService.executeCommand('mssqlCluster.task.newNotebook', this.connectionManagementService.connectionInfo.connectionProfile);
+	}
+
+	public manageExtensions(): void {
+		// TODO: implement what to do here
+		console.error('clicked manage extensions');
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/dashboard/browser/core/dashboardPage.component.ts]: Just have this no-op for now. (and I'd suggest just hiding it until we have it doing something so people aren't confused)

File: src/sql/workbench/contrib/dashboard/browser/core/dashboardPage.css [hunk 6]
```diff
@@ -23,3 +23,8 @@ dashboard-page .dashboard-panel .tab-header .action-item .action-label.pin {
 dashboard-page .dashboard-panel .tab-header .action-item .action-label.close {
 	padding: 5px;
 }
+
+dashboard-page .actions-container .action-item .dashboard-toolbar-item {
+	padding-left: 20px;
+	margin-right: 10px;
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/dashboard/browser/core/dashboardPage.css]: Why are you mixing padding and margin here?

File: src/sql/workbench/contrib/dashboard/browser/pages/databaseDashboardPage.component.ts [hunk 7]
```diff
@@ -57,4 +65,17 @@ export class DatabaseDashboardPage extends DashboardPage implements OnInit {
 		this.init();
 		this._breadcrumbService.setBreadcrumbs(BreadcrumbClass.DatabasePage);
 	}
+
+	protected createTaskbar(parentElement: HTMLElement): void {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/dashboard/browser/pages/databaseDashboardPage.component.ts]: This seems like a better place to add the actions in - not in the base class. 

Also have you thought about how extensions would contribute actions to the taskbar? Ideally the JSON contribution for the tasks-widget would stay basically the same.

File: src/sql/media/icons/backup-toolbar.svg (hunk 1 above)
Comment (kisantia) [reply src/sql/media/icons/backup-toolbar.svg]: Yup. Raden will get back to me after checking with visual designers.

File: src/sql/workbench/contrib/dashboard/browser/core/dashboardPage.component.ts [hunk 8]
```diff
@@ -56,6 +65,16 @@ export abstract class DashboardPage extends AngularDisposable implements IConfig
 
 	@ViewChildren(TabChild) private _tabs: QueryList<DashboardTab>;
 	@ViewChild(PanelComponent) private _panel: PanelComponent;
+	@ViewChild('taskBar', { read: ElementRef }) private taskContainer: ElementRef;
+	protected taskbar: Taskbar;
+
+	// actions
+	protected _editAction: EditDashboardAction;
```
Comment (kburtram) [inline src/sql/workbench/contrib/dashboard/browser/core/dashboardPage.component.ts]: this doesn't seem like a scalable way to add toolbar action.  It seems like this should be contribution driven to allow easier customization.

File: src/sql/workbench/contrib/dashboard/browser/core/dashboardPage.component.ts (hunk 3 above)
Comment (kburtram) [reply src/sql/workbench/contrib/dashboard/browser/core/dashboardPage.component.ts]: yes, the toolbar needs to be able to have different items per page.

**Comment by kisantia** [comment]
I'll make changes so that actions can be contributed to the toolbar similar to tasks-widget. Will send out a new PR after reviewing the design.

**Pull request closed without merging**